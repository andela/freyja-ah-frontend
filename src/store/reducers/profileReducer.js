import * as actionTypes from '../actions/types';

export const initialState = {
  error: null,
  data: {},
  isLoading: false,
};

export default (state = initialState, action) => {
  switch (action.type) {
    case actionTypes.GET_PROFILE_SUCCESS:
      return {
        ...state,
        data: action.data,
      };
    case actionTypes.GET_PROFILE_FAILED:
      return {
        ...state,
        error: action.error,
      };
    case actionTypes.UPLOAD_START:
      return {
        ...state,
        isLoading: true,
      };
    case actionTypes.UPLOAD_FAILED:
      return {
        ...state,
        isLoading: false,
        error: action.error,
      };
    case actionTypes.UPLOAD_SUCCESS:
      return {
        ...state,
        isLoading: false,
        data: {
          ...state.data,
          image: action.imageUrl,
        },
      };
    case actionTypes.UPDATE_PROFILE_FAILED:
      return {
        ...state,
        error: action.error,
      };
    default:
      return state;
  }
};
