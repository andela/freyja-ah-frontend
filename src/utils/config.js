const image = 'https://cdn.umnbootcamp.com/wp-content/uploads/sites/93/2018/03/placeholder-person.png';
const baseUrl = 'https://freyja-ah-backend.herokuapp.com/api';

export {
  image,
  baseUrl,
};
