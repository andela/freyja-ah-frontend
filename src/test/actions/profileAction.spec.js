import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import nock from 'nock';
import {
  getProfile,
  getProfileSuccess,
  uploadSuccess,
  uploadFailed,
  updateProfileFailed,
  updateProfile,
} from '../../store/actions/profileActions';
import { testBaseUrl } from '../config/testConfig';
import profileReducer, { initialState } from '../../store/reducers/profileReducer';

const token = 'token';
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

const mockProfile = {
  data: {
    id: 1,
    phoneNumber: 3223232,
    dateOfBirth: '12/12/12',
    isEmployed: true,
    yrsOfExperience: 4,
    industry: 'Andela',
    image: 'http://',
    isCertified: false,
  },
};

describe('action creators', () => {
  const store = mockStore({});

  it('calls PROFILE_SUCCESS when getProfile is successful', (done) => {
    store.dispatch(getProfileSuccess(mockProfile));
    expect(store.getActions()).toMatchSnapshot();
    done();
  });
});
describe('async actions', () => {
  let store;
  beforeEach(() => {
    store = mockStore({});
  });
  afterEach(() => {
    // clear all HTTP mocks after each test
    nock.cleanAll();
  });
  it('should create getProfileSuccess when profile has been gotten', (done) => {
    nock(`${testBaseUrl}`)
      .get('/api/profiles/4') // Route to catch and mock
      .reply(200, {
        user: {
          userId: 4,
          id: 1,
          firstName: 'John',
          lastName: 'Grisham',
          userName: 'Tyak',
          profile: {
            age: 40,
            phoneNumber: 7689567890,
            bio: 'Help me',
          },
        },
      });
    store.dispatch(getProfile(4, token)).then(() => {
      expect(store.getActions()).toMatchSnapshot();
      done();
    });
  });
  it('should dispatch getProfileFailed when getProfile fails', (done) => {
    nock(`${testBaseUrl}`)
      .get('/api/profiles/99') // Route to catch and mock
      .reply(404, { error: 'user does not exist' });
    store.dispatch(getProfile(4, token)).then(() => {
      expect(store.getActions()).toMatchSnapshot();
      done();
    });
  });
  it('should dispatch updateProfile', (done) => {
    nock(`${testBaseUrl}`)
      .put('/api/profiles', { industry: 'Andela' })
      .reply(201, { profile: { userId: 45 } });
    store.dispatch(updateProfile({ industry: 'Andela' })).then(() => {
      expect(store.getActions()).toMatchSnapshot();
      done();
    });
  });
  it('should dispatch update profile failed', (done) => {
    nock(`${testBaseUrl}`)
      .put('/api/profiles', { dateOfBirth: '10-12' })
      .reply(422, { error: 'invalid date of birth' });
    return store.dispatch(updateProfile({ dateOfBirth: '10-12' })).then(() => {
      expect(store.getActions()).toMatchSnapshot();
      done();
    });
  });
});

describe('profile actions', () => {
  const store = mockStore({});
  it('dispatch UPLOAD_SUCCESS with image url', (done) => {
    store.dispatch(uploadSuccess('http//'));
    expect(store.getActions()).toMatchSnapshot();
    done();
  });
  it('dispatch UPLOAD_FAILED with error object', (done) => {
    store.dispatch(uploadFailed({ error: 'unable to upload image' }));
    expect(store.getActions()).toMatchSnapshot();
    done();
  });
  it('dispatch updateProfileFailed with error object', (done) => {
    store.dispatch(updateProfileFailed({ error: 'invalie information' }));
    expect(store.getActions()).toMatchSnapshot();
    done();
  });
});

describe('profile reducer', () => {
  it('should test the initial state', () => {
    expect(profileReducer(undefined, {})).toMatchSnapshot();
  });
  it('should test the state for update when get profile fails', () => {
    expect(profileReducer(initialState, { type: 'GET_PROFILE_FAILED' })).toMatchSnapshot();
  });
  it('should test the state for update when get profile success', () => {
    expect(profileReducer(initialState, { type: 'GET_PROFILE_SUCCESS' })).toMatchSnapshot();
  });
  it('should test the state for update when upload image success', (done) => {
    expect(profileReducer(initialState, { type: 'UPLOAD_SUCCESS' })).toMatchSnapshot();
    done();
  });
  it('should test the state for update when upload image failed', (done) => {
    expect(profileReducer(initialState, { type: 'UPLOAD_FAILED' })).toMatchSnapshot();
    done();
  });
  it('should test the state for update when upload image failed', (done) => {
    expect(profileReducer(initialState, { type: 'UPLOAD_START' })).toMatchSnapshot();
    done();
  });
});
