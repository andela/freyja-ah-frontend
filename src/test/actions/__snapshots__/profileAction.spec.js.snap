// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action creators calls PROFILE_SUCCESS when getProfile is successful 1`] = `
Array [
  Object {
    "data": Object {
      "data": Object {
        "dateOfBirth": "12/12/12",
        "id": 1,
        "image": "http://",
        "industry": "Andela",
        "isCertified": false,
        "isEmployed": true,
        "phoneNumber": 3223232,
        "yrsOfExperience": 4,
      },
    },
    "type": "GET_PROFILE_SUCCESS",
  },
]
`;

exports[`async actions should create getProfileSuccess when profile has been gotten 1`] = `
Array [
  Object {
    "data": Object {
      "DOB": null,
      "bio": "Help me",
      "email": undefined,
      "image": undefined,
      "industry": undefined,
      "isCertified": undefined,
      "isEnrolled": undefined,
      "name": "John Grisham",
      "phoneNumber": 7689567890,
      "socialMedia": Array [
        Object {
          "name": "facebook",
          "url": undefined,
        },
        Object {
          "name": "twitter",
          "url": undefined,
        },
        Object {
          "name": "linkedin",
          "url": undefined,
        },
      ],
      "username": "Tyak",
      "yrsOfExperience": undefined,
    },
    "type": "GET_PROFILE_SUCCESS",
  },
]
`;

exports[`async actions should dispatch getProfileFailed when getProfile fails 1`] = `
Array [
  Object {
    "error": [Error: Nock: No match for request {
  "method": "GET",
  "url": "https://freyja-ah-backend.herokuapp.com/api/profiles/4",
  "headers": {
    "accept": "application/json, text/plain, */*",
    "authorization": {},
    "user-agent": "axios/0.19.0"
  }
}],
    "type": "GET_PROFILE_FAILED",
  },
]
`;

exports[`async actions should dispatch update profile failed 1`] = `
Array [
  Object {
    "error": "invalid date of birth",
    "type": "UPDATE_PROFILE_FAILED",
  },
]
`;

exports[`async actions should dispatch updateProfile 1`] = `
Array [
  Object {
    "error": [Error: Nock: No match for request {
  "method": "GET",
  "url": "https://freyja-ah-backend.herokuapp.com/api/profiles/45",
  "headers": {
    "accept": "application/json, text/plain, */*",
    "authorization": {},
    "user-agent": "axios/0.19.0"
  }
}],
    "type": "GET_PROFILE_FAILED",
  },
]
`;

exports[`profile actions dispatch UPLOAD_FAILED with error object 1`] = `
Array [
  Object {
    "imageUrl": "http//",
    "type": "UPLOAD_SUCCESS",
  },
  Object {
    "error": Object {
      "error": "unable to upload image",
    },
    "type": "UPLOAD_FAILED",
  },
]
`;

exports[`profile actions dispatch UPLOAD_SUCCESS with image url 1`] = `
Array [
  Object {
    "imageUrl": "http//",
    "type": "UPLOAD_SUCCESS",
  },
]
`;

exports[`profile actions dispatch updateProfileFailed with error object 1`] = `
Array [
  Object {
    "imageUrl": "http//",
    "type": "UPLOAD_SUCCESS",
  },
  Object {
    "error": Object {
      "error": "unable to upload image",
    },
    "type": "UPLOAD_FAILED",
  },
  Object {
    "error": Object {
      "error": "invalie information",
    },
    "type": "UPDATE_PROFILE_FAILED",
  },
]
`;

exports[`profile reducer should test the initial state 1`] = `
Object {
  "data": Object {},
  "error": null,
  "isLoading": false,
}
`;

exports[`profile reducer should test the state for update when get profile fails 1`] = `
Object {
  "data": Object {},
  "error": undefined,
  "isLoading": false,
}
`;

exports[`profile reducer should test the state for update when get profile success 1`] = `
Object {
  "data": undefined,
  "error": null,
  "isLoading": false,
}
`;

exports[`profile reducer should test the state for update when upload image failed 1`] = `
Object {
  "data": Object {},
  "error": undefined,
  "isLoading": false,
}
`;

exports[`profile reducer should test the state for update when upload image failed 2`] = `
Object {
  "data": Object {},
  "error": null,
  "isLoading": true,
}
`;

exports[`profile reducer should test the state for update when upload image success 1`] = `
Object {
  "data": Object {
    "image": undefined,
  },
  "error": null,
  "isLoading": false,
}
`;
